### PR TITLE
drivers: bluetooth: Add Ambiq HCI driver for Apollo4 Blue Plus

### PR DESCRIPTION
**Pull request by aaronyegx**
This PR implements the SPI based HCI driver for Ambiq Apollo4 Blue Plus. We will use HCI-SPI to communicate between the Zephyr Bluetooth host stack and Ambiq Apollo4 Blue Plus internal Bluetooth controller.
The Bluetooth low energy controller inside of Apollo4 Blue Plus is another die running Cortex-M0 core. It works as SPI slave, while Apollo4 Blue Plus host (Cortex-M4) works as SPI master. There are only SPI, interrupt GPIO, clock request GPIO, reset GPIO, XO32MHz and XO32kHz connecting internally between controller and host. The host provides the clock source to controller. No pin of controller die is exposed on the Apollo4 Blue Plus soc.

This PR should be merged after the hal_ambiq PR https://github.com/zephyrproject-rtos/hal_ambiq/pull/9 merged.

Test passed on the following samples:
  -samples/bluetooth/peripheral
  -samples/bluetooth/peripheral_hr
  -samples/bluetooth/peripheral_ht
  -samples/bluetooth/peripheral_esp
  -samples/bluetooth/peripheral_dis
  -samples/bluetooth/peripheral_gatt_write
  -samples/bluetooth/peripheral_accept_list
  -samples/bluetooth/peripheral_sc_only
  -samples/bluetooth/central
  -samples/bluetooth/central_hr
  -samples/bluetooth/central_ht
  -samples/bluetooth/central_gatt_write
  -samples/bluetooth/central_past
  -samples/bluetooth/broadcaster
  -samples/bluetooth/broadcaster_multiple
  -samples/bluetooth/beacon
  -samples/bluetooth/observer
  -samples/bluetooth/direct_adv
  -samples/bluetooth/scan_adv
  -samples/bluetooth/mtu_update
  -samples/bluetooth/direction_finding_central
  -samples/bluetooth/direction_finding_peripheral
  -samples/bluetooth/periodic_adv
  -samples/bluetooth/periodic_sync
  -samples/bluetooth/eddystone
  -samples/bluetooth/ibeacon
  -samples/bluetooth/hci_uart
  -tests/bluetooth/init

### CODE REVIEW

**Review by jhedberg — CHANGES_REQUESTED** [review]
Thanks for the submission! A few initial comments inline.

File: drivers/bluetooth/hci/hci_ambiq.c
Comment (aaronyegx) [reply drivers/bluetooth/hci/hci_ambiq.c:107]: Hi @jhedberg , thanks for your comments. Yes, we are integrating Apollo3 Blue Plus which uses SPI for the HCI as well, so I just try to reserve the condition then I can add Apollo3 Blue Plus support soon. Do you think we still need to remove them at this moment?

**Comment by jhedberg** [comment]
Btw, now that we're starting to have multiple SPI HCI drivers, it makes the existence of `drivers/bluetooth/hci/spi.c` potentially confusing. As there is no standard SPI HCI transport specification (i.e. it's not part of the Bluetooth Core Specification) we should probably not have a driver whose name implies that it's "the standard". Instead, the names of the SPI HCI drivers should likely reflect the specific controller implementation or platform that they support. `spi.c` could perhaps contain the shared parts that all SPI-based HCI drivers need (assuming there's any substantial amount of code that can be shared between them).

@HoZHel @msmttchr @arbrauns I'd appreciate your thoughts on this.

File: drivers/bluetooth/hci/hci_ambiq.c
Comment (RichardSWheatley) [reply drivers/bluetooth/hci/hci_ambiq.c:225]: @jhedberg
We have other apollo3/4 variants they are coming.
That is why this should stay.

File: drivers/bluetooth/hci/hci_ambiq.c
Comment (aaronyegx) [reply drivers/bluetooth/hci/hci_ambiq.c]: I removed the `#if `condition for `"CONFIG_BT_HCI_SETUP"` in the C file.

File: drivers/bluetooth/hci/hci_ambiq.c
Comment (aaronyegx) [reply drivers/bluetooth/hci/hci_ambiq.c]: I removed the `#if defined(CONFIG_SOC_APOLLO4P_BLUE)` at this moment. Will add the adaptable condition when I add Apollo3 Blue support further.

**Comment by msmttchr** [comment]
> Hi @msmttchr @jhedberg , thanks for pointing this. I believe it will be good if we can have one common file to include the shared parts (like spi, bt, etc) for SPI based HCI driver. Ambiq also has different Bluetooth SoCs using different SPI HCI protocol. For now, I split the Ambiq HCI driver to two parts in this PR: `hci_ambiq.c` (maybe similar with the `spi_zephyr.c` we talk about) to include the shared parts, while `apollox_blue.c` to include to specific SPI protocol for Ambiq Apollox Blue soc. If we have a plan to refactor the HCI SPI driver for all chips in the future, I'm happy to adapt the Ambiq one.

@aaronyegx, I think your “draft” idea on how to make a future split is interesting. Also, I forgot to mention in my comment that it could be useful also to have a look at PR #65172 by @HoZHel  for a proposed implementation of ST v1 and v2 protocols.

**Comment by aaronyegx** [comment]
> > Hi @msmttchr @jhedberg , thanks for pointing this. I believe it will be good if we can have one common file to include the shared parts (like spi, bt, etc) for SPI based HCI driver. Ambiq also has different Bluetooth SoCs using different SPI HCI protocol. For now, I split the Ambiq HCI driver to two parts in this PR: `hci_ambiq.c` (maybe similar with the `spi_zephyr.c` we talk about) to include the shared parts, while `apollox_blue.c` to include to specific SPI protocol for Ambiq Apollox Blue soc. If we have a plan to refactor the HCI SPI driver for all chips in the future, I'm happy to adapt the Ambiq one.
> 
> @aaronyegx, I think your “draft” idea on how to make a future split is interesting. Also, I forgot to mention in my comment that it could be useful also to have a look at PR #65172 by @HoZHel for a proposed implementation of ST v1 and v2 protocols.

Hi @msmttchr , thanks for sharing the link. When I started to implement the HCI driver for Ambiq devices, I referred to the drivers/bluetooth/hci/spi.c and understood it was some difficult to reuse that one for Ambiq SoC, so I created a new one. Ambiq Apollo3/4 Blue Plus use different SPI protocols for HCI but may be able to use same dts binding. I have split the Ambiq HCI driver to two parts/files in the latest push of this PR, you can talk a look if possible. As commented by @jhedberg previously, the file may become large if we put all the devices support/protocol together. So, I split them.

File: drivers/bluetooth/hci/apollox_blue.c
Comment (msmttchr) [inline drivers/bluetooth/hci/apollox_blue.c:284]: I understand that in very early stage, the device needed some configuration commands.
In this case, the driver calls higher level software api to self inject a command and handle reply.
Even though this is also used in the `spi.c` driver which inspired this code, I think this prevents compiling sample applications like `hci_uart` or similar that use hci raw access which does not provide support for api like `bt_hci_cmd_create`.
A possible solution for this problem is to use low level api provided by the driver itself (eg.  `bt_apollo_spi_send`), hence removing dependencies on `hci_core.c` file.

Please take this comment as point to discuss and not as request or recommendation for changes.

**Comment by aaronyegx** [comment]
Hi @jhedberg , if possible, could you help revisit this PR? We Ambiq are close to deadline of some projects, and we would like to try our best to address the comments and get this feature done as soon as possible. Thanks for your understanding.

File: drivers/bluetooth/hci/apollox_blue.c
Comment (jhedberg) [reply drivers/bluetooth/hci/apollox_blue.c:321]: In the case of building Zephyr as a pure controller + HCI transport with the help of `BT_HCI_RAW` I'd expect the host-side of the HCI transport (which could e.g. be Linux) to take over the responsibility of this kind of initialisation.

Comment (jhedberg) [reply drivers/bluetooth/hci/apollox_blue.c:321]: That said, we could play with the idea of extracting out HCI command handling code from `hci_core.c` into a kind of library, which could optionally be included even in `BT_HCI_RAW` builds, however my initial feeling is that this is kind of violating what `BT_HCI_RAW` was created for, i.e. a simple and "dumb" pipe to the controller, so that another host can talk to it (and potentially also do early HCI init) instead of Zephyr.

File: drivers/bluetooth/hci/hci_ambiq.c
Comment (jhedberg) [inline drivers/bluetooth/hci/hci_ambiq.c:78]: Wouldn't it be cleaner to have a proper header file for these?

File: drivers/bluetooth/hci/apollox_blue.c
Comment (msmttchr) [reply drivers/bluetooth/hci/apollox_blue.c:321]: @jhedberg, I think it is a bit of grey area here, at least for me :). If we assume that Linux Host should now that a special command(s) needs to be sent to the device before starting operation, question is:

- How is Linux Host supposed to know about this?
- On the other hand, why Zephyr Host is not aware of this?
- What about connecting a chip running hci_uart (HCI raw) and a chip running Zephyr host ? [Obviously the Zephyr host will not know that some config commands are needed].

I was thinking that we can assume that some devices, normally designed for custom host implementation, requires some "setup" commands before starting standard operation. It is quite common to assume that this "setup" commands are sent using the standard hardware interface (SPI in our case) to avoid adding other physical interfaces between the two chips, so I would consider that during the setup phase, the device is not fully operational and so the regular hci command (even vendor specific) are not guaranteed to work.

If this agreed, then we can think of a setup phase (already present) where all these special configuration commands are sent, without relying on higher level software. After the setup phase, then regular HCI command (standard or vendor specific) needs to be supported.

In summary, we can simply forbid to use things like `bt_hci_cmd_send_sync` or similar during the setup phase or in general in all the driver.

Comment (msmttchr) [reply drivers/bluetooth/hci/apollox_blue.c:321]: > Hi @msmttchr , thanks for your comment. Yes we always need samples for hci raw access for DTM or other purpose. It will have `bt_hci_cmd_create` compiling issue for that samples. I add `CONFIG_BT_HCI_RAW` condition here to add support for hci raw access. I would like to keep using `bt_hci_cmd_create` and `bt_hci_cmd_send_sync` for the vendor specific commands in non-raw case to make sure they are sent successfully since there is expected semaphore inside of these APIs to control.

@aaronyegx, I think adding `CONFIG_BT_HCI_RAW` condition is a good short-term approach to address the issue.

File: drivers/bluetooth/hci/hci_ambiq.c
Comment (aaronyegx) [reply drivers/bluetooth/hci/hci_ambiq.c]: I added a header file in hal_ambiq to declare these functions to make the HCI driver cleaner.

File: drivers/bluetooth/hci/apollox_blue.c
Comment (aaronyegx) [reply drivers/bluetooth/hci/apollox_blue.c:320]: @jhedberg , in some cases the "host" is just one serial tool or test machine which is requested to send the hci sequences and it has less knowledge about the controller running in the platform. It may be unavailable to do any initialization for the hci transport of Zephyr platform. The Bluetooth stack hci layer defines hci_core.c and hci_raw.c for general hci and hci raw access APIs, the hci driver here above the hci stack should be the best to decide whether the hci core or raw APIs to be executed in current framework.

File: drivers/bluetooth/hci/hci_ambiq.c
Comment (jhedberg) [reply drivers/bluetooth/hci/hci_ambiq.c]: > I added a header file in hal_ambiq to declare these functions to make the HCI driver cleaner.

Aren't all of these APIs exposed implemented by `apollox_blue.c`? The right place for the header file would therefore be here in the same directory, not in the HAL tree.

Comment (aaronyegx) [reply drivers/bluetooth/hci/hci_ambiq.c]: > Aren't all of these APIs implemented by `apollox_blue.c`? The right place for the header file would therefore be here in the same directory, not in the HAL tree.

I put the header file `apollox_blue.h` in the same directory with `apollox_blue.c`, thank you.

**Review by jhedberg — DISMISSED** [review]

**Comment by aaronyegx** [comment]
Hi @jhedberg , we merged the hal_ambiq PR and updated the revision in west.yml accordingly. Could you reapprove this? Thank you.

**Review by jhedberg — APPROVED** [review]

**Comment by aaronyegx** [comment]
Hi @fkokosinski @tgorochowik , this PR enables the Bluetooth/HCI feature for Ambiq Apollo4 Blue Plus and incoming other Bluetooth series soc of Ambiq platform. If possible, could you also take a look at this? Thanks very much.

File: drivers/bluetooth/hci/apollox_blue.c
Comment (msmttchr) [reply drivers/bluetooth/hci/apollox_blue.c:319]: @jhedberg, thanks for the clarification, I think I understand a little bit more 😃 now.
I saw the implementation done by @aaronyegx which could be a good compromise at the end.

**Review by msmttchr — APPROVED** [review]
LGTM. I think now also hci_uart is working

**Comment by aaronyegx** [comment]
Hi @fkokosinski @tgorochowik , we passed most of the adaptable Zephyr Bluetooth samples (refer to the test list in the PR description) with the implemented Ambiq HCI driver of this PR on apollo4p_blue_kxr_evb board. We are trying to release this important feature required by the waiting wide customers for evaluation as soon as possible and getting feedback from them for code quality improvement or bug fix.
If there is no big concern, could you help move forward this, please? Thanks very much for your understanding and support.

**Review by fkokosinski — APPROVED** [review]
Board and dts bindings changes LGTM. :)

**Review by carlescufi — CHANGES_REQUESTED** [review]
Before I merge this, may I ask the following 2 questions:

1. ~Why wasn't the existing `hci_spi.c` reused by extending it?~ I realize now by reading the comments.
2. Why the need to the `apollox_blue.*` files if we have `hci_ambiq.c`? Couldn't we just have a single file with all the code?

**Comment by aaronyegx** [comment]
> Before I merge this, may I ask the following 2 questions:
> 
> 1. ~Why wasn't the existing `hci_spi.c` reused by extending it?~ I realize now by reading the comments.
> 2. Why the need to the `apollox_blue.*` files if we have `hci_ambiq.c`? Couldn't we just have a single file with all the code?

Hi @carlescufi , based on our previous discussion in this PR, we will integrate Apollo3 Blue Plus which also uses SPI for HCI but there are many differences between Apollo4 and Apollo3 in SPI protocol. Using a single file was the initial submission in this PR, but we thought using a single file to include all the contents would make the file too large and less readable/maintainable. So, we decide to create `apollox_blue.c` for specific SPI protocol for the soc, while `hci_ambiq.c` will include the common parts. It is most likely we only need to modify `apollox_blue.c` to enable support for Apollo3 Blue Plus in incoming implementation. Thank you.

**Comment by aaronyegx** [comment]
> > Before I merge this, may I ask the following 2 questions:
> > 
> > 1. ~Why wasn't the existing `hci_spi.c` reused by extending it?~ I realize now by reading the comments.
> > 2. Why the need to the `apollox_blue.*` files if we have `hci_ambiq.c`? Couldn't we just have a single file with all the code?
> 
> Hi @carlescufi , based on our previous discussion in this PR (see [#66227 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/66227#discussion_r1418718420)), we will integrate Apollo3 Blue Plus which also uses SPI for HCI but there are many differences between Apollo4 and Apollo3 in SPI protocol. Using a single file was the initial submission in this PR, but we thought using a single file to include all the contents would make the file too large and less readable/maintainable. So, we decide to create `apollox_blue.c` for specific SPI protocol for the soc, while `hci_ambiq.c` will include the common parts. It is most likely we only need to modify `apollox_blue.c` to enable support for Apollo3 Blue Plus in incoming implementation. Thank you.

Hi @carlescufi , could you help revisit this, please? Thank you.

**Comment by aaronyegx** [comment]
> > > Before I merge this, may I ask the following 2 questions:
> > > 
> > > 1. ~Why wasn't the existing `hci_spi.c` reused by extending it?~ I realize now by reading the comments.
> > > 2. Why the need to the `apollox_blue.*` files if we have `hci_ambiq.c`? Couldn't we just have a single file with all the code?
> > 
> > 
> > Hi @carlescufi , based on our previous discussion in this PR (see [#66227 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/66227#discussion_r1418718420)), we will integrate Apollo3 Blue Plus which also uses SPI for HCI but there are many differences between Apollo4 and Apollo3 in SPI protocol. Using a single file was the initial submission in this PR, but we thought using a single file to include all the contents would make the file too large and less readable/maintainable. So, we decide to create `apollox_blue.c` for specific SPI protocol for the soc, while `hci_ambiq.c` will include the common parts. It is most likely we only need to modify `apollox_blue.c` to enable support for Apollo3 Blue Plus in incoming implementation. Thank you.
> 
> Hi @carlescufi , could you help revisit this, please? Thank you.

Hi @carlescufi , could you look at the answer for your merging question? Sorry I marked some comments resolved to make you miss some information in your first review. I think the comments https://github.com/zephyrproject-rtos/zephyr/pull/66227#discussion_r1418718420, https://github.com/zephyrproject-rtos/zephyr/pull/66227#discussion_r1422338143, https://github.com/zephyrproject-rtos/zephyr/pull/66227#discussion_r1422751870 should clear your confusion. We started to announce Zephyr supports Apollo family SoCs, the Bluetooth is the basic supported feature, which is waiting for merging here. Thank you so much.

**Comment by carlescufi** [comment]
> > Before I merge this, may I ask the following 2 questions:
> > 
> > 1. ~Why wasn't the existing `hci_spi.c` reused by extending it?~ I realize now by reading the comments.
> > 2. Why the need to the `apollox_blue.*` files if we have `hci_ambiq.c`? Couldn't we just have a single file with all the code?
> 
> Hi @carlescufi , based on our previous discussion in this PR (see [#66227 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/66227#discussion_r1418718420)), we will integrate Apollo3 Blue Plus which also uses SPI for HCI but there are many differences between Apollo4 and Apollo3 in SPI protocol. Using a single file was the initial submission in this PR, but we thought using a single file to include all the contents would make the file too large and less readable/maintainable. So, we decide to create `apollox_blue.c` for specific SPI protocol for the soc, while `hci_ambiq.c` will include the common parts. It is most likely we only need to modify `apollox_blue.c` to enable support for Apollo3 Blue Plus in incoming implementation. Thank you.

Thanks for your explanation. Approving now.

**Review by carlescufi — APPROVED** [review]

File: boards/arm/apollo4p_blue_kxr_evb/apollo4p_blue_kxr_evb.dts
Comment (fabiobaltieri) [inline boards/arm/apollo4p_blue_kxr_evb/apollo4p_blue_kxr_evb.dts:104]: Are these wired internally in the chip? If that's the case I think it would make sense to have the nodes declared in `ambiq_apollo4p_blue.dtsi` so that you don't have to redefine the same things on every board using the chip, it's not like you can change it anyway. See `stm32wl.dtsi` with the `stm32wl-subghz-radio` node in the internal spi for example.

Comment (aaronyegx) [reply boards/arm/apollo4p_blue_kxr_evb/apollo4p_blue_kxr_evb.dts:104]: @fabiobaltieri thanks for your following. Will have an independent PR to optimize this.

**Pull request merged**
